test/e2e/gateway_api_test.go

```diff
@@ -0,0 +1,212 @@
+		}
+		// Enable the feature gate for the rest of the test
+		featureGate.Spec.FeatureGateSelection.FeatureSet = configv1.CustomNoUpgrade
+		featureGate.Spec.FeatureGateSelection.CustomNoUpgrade = &configv1.CustomFeatureGates{
+			Enabled: []string{gwapi.FeatureGateName},
+		}
+		if err := kclient.Update(context.TODO(), featureGate); err != nil {
+			t.Fatalf("error enabling feature gate: %v", err)
+		}
+	}
+	// Make sure all the *.gateway.networking.k8s.io CRDs are available
+	for _, crdName := range crdNames {
+		err := getCrd(t, crdName)
+		if err != nil {
+			t.Fatalf("failed to find crd %s: %v", crdName, err)
+		}
+	}
+
+	// Now make sure everything can happen with the Gateway API CRDs since the feature is enabled
+	ok, gateway, gatewayClass := testCrds(t, ns.Name, true)
+	defer func() {
+		// Delete the gateway
+		if err := kclient.Delete(context.TODO(), gateway); err != nil {
+			t.Logf("failed to delete test gateway %s: %v", gateway.Name, err)
+		}
+		// Delete the gateway class
+		if err := kclient.Delete(context.TODO(), gatewayClass); err != nil {
+			t.Logf("failed to delete test gatewayclass %s: %v", gatewayClass.Name, err)
+		}
+	}()
+
+	if !ok {
+		t.Fatalf("feature gate was enabled, but gateway api objects could not be created")
+	}
+
+	t.Logf("successful test of gateway api crds")
+}
+
+func getCrd(t *testing.T, name string) error {
+	t.Helper()
+	crd := &apiextensionsv1.CustomResourceDefinition{}
+	err := wait.PollImmediate(100*time.Millisecond, 1*time.Second, func() (bool, error) {
+		name := types.NamespacedName{"", name}
+		err := kclient.Get(context.TODO(), name, crd)
+		if err != nil {
+			t.Logf("failed to get crd %s: %v", name, err)
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+func testCrds(t *testing.T, ns string, log bool) (success bool, gateway *v1beta1.Gateway, gatewayClass *v1beta1.GatewayClass) {
+	t.Helper()
+	success, gatewaySuccess, gatewayClassSuccess, httpRouteSuccess := false, true, true, true
+
+	// Check if GatewayClass can be created.
+	gatewayClass = buildGatewayClass("openshift-default", "openshift.io/gateway-controller")
+	if err := kclient.Create(context.TODO(), gatewayClass); err != nil {
+		gatewayClassSuccess = false
+		if log {
+			t.Logf("error creating gateway class: %v", err)
+		}
+	}
+
+	// Check if Gateway can be created.
+	gateway = buildGateway("test-gateway", "openshift-ingress", gatewayClass.Name)
+	if err := kclient.Create(context.TODO(), gateway); err != nil {
+		gatewaySuccess = false
+		if log {
+			t.Logf("error creating gateway: %v", err)
+		}
+	}
+
+	// Check if HTTPRoute can be created.
+	httpRoute := buildHTTPRoute("test-httproute", ns, gateway.Name, "openshift-ingress", "test-hostname.example.com", "test-app", 8080)
+	if err := kclient.Create(context.TODO(), httpRoute); err != nil {
+		httpRouteSuccess = false
+		if log {
+			t.Logf("error creating http route: %v", err)
+		}
+	}
```

Comment:
Would it make sense also to check the HTTPRoute's status to make sure Istio reconciled it and associated it with the gateway?

Reply:
I can do that once we know everything is working properly in OSSM.  The only thing I know for sure is working properly is in this repo, so I'm only testing what c-i-o should be doing.